pkg/lib/web/generated_asset.go

```diff
@@ -0,0 +1,209 @@
+
+	return result, nil
+}
+
+func (d *GeneratedAssetDescriptor) watch() {
+	for {
+		select {
+		case event, ok := <-d.watcher.Events:
+			if !ok {
+				return
+			}
+			if event.Op&fsnotify.Write != fsnotify.Write {
+				break
+			}
+			d.Logger.
+				WithField("file", event.Name).
+				Info("change detected, reloading...")
+
+			if err := d.reload(); err != nil {
+				d.Logger.
+					WithError(err).
+					WithField("file", event.Name).
+					Error("reload failed")
+			}
+
+		case err, ok := <-d.watcher.Errors:
+			if !ok {
+				return
+			}
+			d.Logger.WithError(err).Fatal("Watcher failed")
+		}
+	}
+}
+
+func (d *GeneratedAssetDescriptor) reload() error {
+	newManifest, err := LoadManifest()
+	if err != nil {
+		return err
+	}
+
+	manifestCtx := &ManifestContext{
+		content: newManifest,
+	}
+	d.manifest.Store(manifestCtx)
+	return nil
+}
+
+func (d *GeneratedAssetDescriptor) MatchResource(resourcePath string) (*resource.Match, bool) {
+	manifest := d.manifest.Load().(*ManifestContext).content
+
+	key := strings.TrimPrefix(resourcePath, StaticAssetResourcePrefix)
+	if IsSourceMapPath(key) {
+		key = strings.TrimSuffix(key, ".map")
+	}
+
+	if _, ok := manifest[key]; ok {
+		return &resource.Match{}, true
+	}
+	return nil, false
+}
+
+func (d *GeneratedAssetDescriptor) FindResources(fs resource.Fs) ([]resource.Location, error) {
+	manifest := d.manifest.Load().(*ManifestContext).content
+	locations := make([]resource.Location, 0)
+
+	for _, value := range manifest {
+		location := resource.Location{
+			Fs:   fs,
+			Path: fmt.Sprintf("static/%s", value),
+		}
+		_, err := resource.ReadLocation(location)
+
+		if os.IsNotExist(err) || err != nil {
+			continue
+		}
+		locations = append(locations, location)
+
+		locationMap := resource.Location{
+			Fs:   fs,
+			Path: fmt.Sprintf("static/%s.map", value),
+		}
+		_, err = resource.ReadLocation(locationMap)
+
+		if os.IsNotExist(err) || err != nil {
+			continue
+		}
+		locations = append(locations, locationMap)
+	}
+	return locations, nil
+}
+
+func (d *GeneratedAssetDescriptor) ViewResources(resources []resource.ResourceFile, rawView resource.View) (interface{}, error) {
+	switch rawView.(type) {
+	case resource.AppFileView:
+		return d.viewResources(resources, rawView)
+	case resource.EffectiveFileView:
+		return d.viewResources(resources, rawView)
+	case resource.EffectiveResourceView:
+		return d.viewResources(resources, rawView)
+	case resource.ValidateResourceView:
+		return d.viewResources(resources, rawView)
+	default:
+		return nil, fmt.Errorf("unsupported view: %T", rawView)
+	}
+}
+
+func (d *GeneratedAssetDescriptor) viewResources(resources []resource.ResourceFile, rawView resource.View) ([]byte, error) {
+	if len(resources) == 0 {
+		return nil, resource.ErrResourceNotFound
+	}
+
+	manifest := d.manifest.Load().(*ManifestContext).content
+	for _, r := range resources {
+		key := strings.TrimPrefix(rawView.(resource.EffectiveFile).Path, StaticAssetResourcePrefix)
```

Comment:
This cast will crash if rawView is not resource.EffectiveFile. If we do not know how to handle other view, return resource.ErrResourceNotFound